api/test_lambda_function.py

```diff
@@ -0,0 +1,65 @@
+from lambda_function import lambda_handler, table
+from unittest.mock import patch
+import json
+
+def test_add_stamps_with_no_matching_record_in_DB():
+    event = {
+        'requestType': 'addStamps',
+        'storeId': '12345',
+        'customerId': '54312',
+        'numberOfStamps': 1
+    }
+
+    expected = {
+    'statusCode': 200,
+    'body': json.dumps('Sucessfully added %s stamps for StoreId: %s and CustomerId: %s' % (event['numberOfStamps'], event['storeId'], event['customerId']))
+    }
+
+    expected_arguments = {
+    	'StoreId': event['storeId'],
+    	'CustomerId': event['customerId'],
+    	'Stamps': event['numberOfStamps']
+    }
+
+    with patch.object(table, 'put_item', return_value={}) as mock_put_item, patch.object(table, 'get_item', return_value={}) as mock_get_item:
+        result = lambda_handler(event, None)
+        mock_put_item.assert_called_once()
+        mock_put_item.assert_called_with(Item=expected_arguments)
+        assert result == expected
+
+def test_add_stamps_with_matching_record_in_DB():
+    event = {
+        'requestType': 'addStamps',
+        'storeId': '12345',
+        'customerId': '54312',
+        'numberOfStamps': 1
+    }
+
+    expected = {
+    'statusCode': 200,
+    'body': json.dumps('Sucessfully added %s stamps for StoreId: %s and CustomerId: %s' % (event['numberOfStamps'], event['storeId'], event['customerId']))
+    }
+
+    get_item_response = {
+    	'Item' : {
+	    	'StoreId': event['storeId'],
+	    	'CustomerId': event['customerId'],
+	   		'Stamps': 1
+	    }
+    }
+
+    expected_key = {
+    	'StoreId': event['storeId'],
+    	'CustomerId': event['customerId']
+    }
+    expected_update_expression = "set Stamps = Stamps + :val"
+    expected_expression_attribute_values = {
+    	':val': event['numberOfStamps']
+    }
+    expected_return_values = "UPDATED_NEW"
+
+    with patch.object(table, 'update_item', return_value={}) as mock_update_item, patch.object(table, 'get_item', return_value=get_item_response) as mock_get_item:
```

Comment:
What does this part do?

Reply:
Unit tests should never change the state of the database and should not depend on the current state of the database. If we call the function as is, then table.get_item and table.update_item will go to the database. So we need a way of testing the function without actually going to the database. This is done by adding mocks - the mock acts as a substitute to the call table.get_item and table.update_item where we can check that the function is called without actually having it execute. 

You can read more about unit testing here - https://towardsdatascience.com/unit-testing-with-mocking-in-10-minutes-e28feb7e530
The last part talks about mocking but it might be worth reading all parts to understand unit testing